### DISCUSSION TITLE
Hasil RNN overfit

### DISCUSSION THREAD

**Discussion by DeanAlexander27** [teamalgoritma/community#586] [Q&A] [answered]
Dear Ka Instruktur
saya cukup kesulitan mengerjakan LBB DSS Pytorch
yang dimana menggunakan data sebagai berikut :
https://www.kaggle.com/datasets/census/population-time-series-data

saya mengalami kesulitan ketika menyederhanakan arsitektur sebagai solusi untuk mengatasi overfitting
![image](https://user-images.githubusercontent.com/123348110/223166845-eb97118c-d8c5-4393-9012-e2b8963f168f.png)

apakah ada solusi untuk seperti ini ?

epochs = 30
train_cost, test_cost = [], []

for i in range(epochs):
    # training
    model.train()
    cost = 0
    for feature, target in train_loader: # iterasi untuk minibatch
        # feedforward
        output, hidden = model.forward(feature, None)
        # hitung loss
        loss = criterion(output, target)
        # backpropagation
        loss.backward()
        # update weight
        optimizer.step()
        # reset optimizer (kasus khusus pada pytorch)
        optimizer.zero_grad()
        
        # perhitungan loss total per minibatch (dikali jumlah data agar bisa dibagi dengan total data di akhir)
        cost += loss.item() * feature.shape[0] 
        # .item() buat ambil skalar nilai 
    train_cost.append(cost / len(train_dataset)) # cost (MSE) per epoch dicatat
    
    # evaluasi
    with torch.no_grad(): # memastikan tidak terjadi perhitungan gradient
        model.eval()
        cost = 0
        for feature, target in test_loader:
            # feedforward
            output, hidden = model.forward(feature, None)
            # hitung loss sebagai pembanding
            loss = criterion(output, target)
            # data test ga backpropagation dan update weight nanti leakage
            
            # perhitungan loss total per minibatch (dikali jumlah data agar bisa dibagi dengan rata2)
            cost += loss.item() * feature.shape[0]
        test_cost.append(cost / len(test_dataset)) # cost (MSE) per epoch dicatat

    print(f'\rEpoch: {i+1}/{epochs} | train_cost: {train_cost[-1]:4f} | test_cost: {test_cost[-1]:4f}', end=' ')

**Comment by idoali** [comment]
Terima kasih atas pertanyaannya pak

Apa boleh kami minta code lengkap atau file jupyter notebooknya pak?

**Reply by DeanAlexander27** [reply]
Halo Ka Ido... baik ka.. apa ada media share yang bisa saya kirim dalam ipynb ?

**Reply by idoali** [reply]
Bisa langsung dikirim disini pak. Caranya tinggal di drag saja.

**Reply by DeanAlexander27** [reply]
Baik ka sudah saya create dalam zip beserta datasetnya 
[LBB_Dean_Alexander.zip](https://github.com/teamalgoritma/community/files/10917602/LBB_Dean_Alexander.zip)

terima kasih, sehat selalu

**Reply by idoali** [reply]
Mohon maaf pak baru membalas lagi

Ada beberapa kesalahan disini pak. Data yang bapak punya nilainya terlalu besar, sehingga kesalahan sedikit saja akan menunjukkan error yang sangat besar. 

Dapat dilihat juga bahwa data bapak memiliki trend meningkat setiap waktunya. Lalu ketika bapak membagi data menjadi data train dan data test, data test adalah data terbaru. Karena data test adalah data yang terbaru, maka nilainya juga jauh lebih besar ketimbang data train. 

Solusinya bisa dicoba tahapan evaluasi tidak dimasukkan ke loop yang sama dengan tahapan training.  

```{python}
for i in range(epochs):
    # training
    model.train()
    cost = 0
    for feature, target in train_loader: # iterasi untuk minibatch
        # feedforward
        output, hidden = model.forward(feature, None)
        # hitung loss
        loss = criterion(output, target)
        # backpropagation
        loss.backward()
        # update weight
        optimizer.step()
        # reset optimizer (kasus khusus pada pytorch)
        optimizer.zero_grad()
        
        # perhitungan loss total per minibatch (dikali jumlah data agar bisa dibagi dengan total data di akhir)
        cost += loss.item() * feature.shape[0] 
        # .item() buat ambil skalar nilai 
    train_cost.append(cost / len(train_dataset)) # cost (MSE) per epoch dicatat
    
# evaluasi
with torch.no_grad(): # memastikan tidak terjadi perhitungan gradient
    model.eval()
    cost = 0
    for feature, target in test_loader:
        # feedforward
        output, hidden = model.forward(feature, None)
        # hitung loss sebagai pembanding
        loss = criterion(output, target)
        # data test ga backpropagation dan update weight nanti leakage

        # perhitungan loss total per minibatch (dikali jumlah data agar bisa dibagi dengan rata2)
        cost += loss.item() * feature.shape[0]
    test_cost.append(cost / len(test_dataset)) # cost (MSE) per epoch dicatat
```

**Reply by DeanAlexander27** [reply]
ahh.. baik Ka Ido... segera saya uji cobakan,, saya cukup curiga juga karena memang sifat data saya sepertinya akumulasi.. karena mengenai data populasi.. apakah itu juga mempengaruhi atau dapat diterima dalam model ini ?

**Reply by idoali** [reply]
Baik pak. Kalau begitu saya minta izin untuk menulis ulang jawabannya dan ditandai. 

Semisal masih bermasalah, bisa juga angkanya di-scale, menggunakan `StandardScaler()` contohnya, agar nantinya angkanya tidak terlalu besar.

**Comment by idoali** [comment, marked as answer]
Mohon maaf pak baru membalas lagi

Ada beberapa kesalahan disini pak. Data yang bapak punya nilainya terlalu besar, sehingga kesalahan sedikit saja akan menunjukkan error yang sangat besar.

Dapat dilihat juga bahwa data bapak memiliki trend meningkat setiap waktunya. Lalu ketika bapak membagi data menjadi data train dan data test, data test adalah data terbaru. Karena data test adalah data yang terbaru, maka nilainya juga jauh lebih besar ketimbang data train.

Solusinya bisa dicoba tahapan evaluasi tidak dimasukkan ke loop yang sama dengan tahapan training.

```{}
for i in range(epochs):
    # training
    model.train()
    cost = 0
    for feature, target in train_loader: # iterasi untuk minibatch
        # feedforward
        output, hidden = model.forward(feature, None)
        # hitung loss
        loss = criterion(output, target)
        # backpropagation
        loss.backward()
        # update weight
        optimizer.step()
        # reset optimizer (kasus khusus pada pytorch)
        optimizer.zero_grad()
        
        # perhitungan loss total per minibatch (dikali jumlah data agar bisa dibagi dengan total data di akhir)
        cost += loss.item() * feature.shape[0] 
        # .item() buat ambil skalar nilai 
    train_cost.append(cost / len(train_dataset)) # cost (MSE) per epoch dicatat
    
# evaluasi
with torch.no_grad(): # memastikan tidak terjadi perhitungan gradient
    model.eval()
    cost = 0
    for feature, target in test_loader:
        # feedforward
        output, hidden = model.forward(feature, None)
        # hitung loss sebagai pembanding
        loss = criterion(output, target)
        # data test ga backpropagation dan update weight nanti leakage

        # perhitungan loss total per minibatch (dikali jumlah data agar bisa dibagi dengan rata2)
        cost += loss.item() * feature.shape[0]
    test_cost.append(cost / len(test_dataset)) # cost (MSE) per epoch dicatat
```

**Reply by DeanAlexander27** [reply]
maaf baru membalas ka Ido.
hasil grafik yang keluar hanya train saja
![image](https://user-images.githubusercontent.com/123348110/225973232-15b963af-8e43-420c-8a65-f4741c97765c.png)

tetapi ketika saya coba masukkan ke dalam loopingnya sudah berjalan seperti berikut
![image](https://user-images.githubusercontent.com/123348110/225973576-21cb0273-d509-44c2-8c6e-7be0ffa0e65e.png)

setelah masuk ke prediksi
hasilnya cukup aneh. karena tidak selaras

![image](https://user-images.githubusercontent.com/123348110/225973661-1da511dd-6920-4770-bf25-af62493fd9c0.png)

[LBB_Dean_Alexander.zip](https://github.com/teamalgoritma/community/files/11004008/LBB_Dean_Alexander.zip)
